File: benchmarks/faastest/arsenal/sleep/openfaas/sleep-py/handler.py
import os
import time
import datetime

def get_current_epoch():
    return int((datetime.datetime.utcnow() - datetime.datetime(1970, 1, 1)).total_seconds() * 1000)

def get_sleep_parameter(event):
    user_input = str(event.query["sleep"])
    if not user_input or not user_input.isdigit() or int(user_input) < 0:
        return {"error": "invalid sleep parameter"}
    return int(user_input)

def run_test(sleep_time):
    time.sleep(sleep_time / 1000.0)

def is_warm():
    is_warm = os.environ.get("warm") == "true"
    os.environ["warm"] = "true"
    return is_warm

def handle(event):
    start = get_current_epoch()
    reused = is_warm()

    sleep_time = get_sleep_parameter(event)
    if type(sleep_time) != int:
        return {
            "statusCode": 200,
            "body": sleep_time
        }

    run_test(sleep_time)
    duration = (get_current_epoch() - start) * 1000000

    return {
        "statusCode": 200,
        "body": {
            "duration": duration,
            "reused": reused
        }
    }

File: benchmarks/faastest/arsenal/sleep/openfaas/sleep-node/handler.js
const wait = ms => new Promise((r, j) => setTimeout(r, ms));

function getDuration(startTime) {
    const end = process.hrtime(startTime);
    return end[1] + (end[0] * 1e9);
}

function getSleep(event) {
    const sleep_time = event.query.sleep ? parseInt(event.query.sleep) : null;
    if (!sleep_time && sleep_time !== 0) {
        return {"error": "invalid sleep parameter"};
    }
    return sleep_time;
}

function getParameters(event) {
    return getSleep(event);
}

async function runTest(sleep_time){
    await wait(sleep_time);
}

function isWarm() {
    const is_warm = process.env.warm ? true : false;
    process.env.warm = true;
    return is_warm;
}

module.exports = async (event, context) => {
    const startTime = process.hrtime();
    const params = getParameters(event);
    if (params.error) {
        return context.status(200).succeed({
            error: params.error,
        });
    }

    await runTest(params);

    const reused = isWarm();
    const duration = getDuration(startTime);

    return context.status(200).succeed({
        reused: reused,
        duration: duration,
    });
}
